internal/cimfs/cim_reader_windows.go

```diff
@@ -0,0 +1,1074 @@
+		}
+		sr.peinit = true
+	}
+	d := int64(0)
+	end := sr.pe.DataLength
+	for _, m := range sr.pemappings {
+		if uint64(m.FileOffset) > off {
+			end = uint64(m.FileOffset)
+			break
+		}
+		d = int64(m.Delta)
+	}
+	return d, end - off, nil
+}
+
+func (cr *CimFsReader) readStream(sr *streamReader, b []byte) (_ int, err error) {
+	n := len(b)
+	rem := sr.stream.Size() - sr.off
+	if uint64(n) > rem {
+		n = int(rem)
+	}
+	if sr.stream.Sparse() {
+		for i := range b[:n] {
+			b[i] = 0
+		}
+		return n, nil
+	}
+	ro := sr.stream.DataOffset
+	off := sr.off
+	if sr.stream.Type() == format.StreamTypePeImage {
+		delta, segrem, err := cr.getPESegment(sr, sr.off)
+		if err != nil {
+			return 0, err
+		}
+		if uint64(n) > segrem {
+			n = int(segrem)
+		}
+		ro = sr.pe.DataOffset
+		off += uint64(delta)
+	}
+	n, err = cr.readOffsetFull(b[:n], ro, off)
+	sr.off += uint64(n)
+	rem -= uint64(n)
+	if err == nil && rem == 0 {
+		err = io.EOF
+	}
+	return n, err
+}
+
+func (f *File) Read(b []byte) (_ int, err error) {
+	defer func() {
+		if err != nil && err != io.EOF {
+			err = &PathError{Cim: f.r.name, Path: f.Name(), Op: "read", Err: err}
+		}
+	}()
+	if f.IsDir() {
+		return 0, ErrIsADirectory
+	}
+	return f.r.readStream(&f.sr, b)
+}
+
+const (
+	ltNameOffSize = 4
+	ltNameLenSize = 2
+	ltSizeOff     = 0
+	ltCountOff    = 4
+	ltEntryOff    = 8
+	fileIDSize    = 4
+	streamSize    = 16
+)
+
+func parseName(b []byte, nos []byte, i int) ([]byte, error) {
+	size := uint32(len(b))
+	no := binary.LittleEndian.Uint32(nos[i*ltNameOffSize:])
+	if no > size-ltNameLenSize {
+		return nil, fmt.Errorf("invalid name offset %d > %d", no, size-ltNameLenSize)
+	}
+	nl := binary.LittleEndian.Uint16(b[no:])
+	if mnl := (size - ltNameLenSize - no) / 2; uint32(nl) > mnl {
+		return nil, fmt.Errorf("invalid name length %d > %d", nl, mnl)
+	}
+	return b[no+ltNameLenSize : no+ltNameLenSize+uint32(nl)*2], nil
+}
+
+// cmpcaseUtf8Utf16LE compares a UTF-8 string with a UTF-16LE encoded byte
+// array, upcasing each rune through the upcase table.
+func cmpcaseUtf8Utf16LE(a string, b []byte, upcase []uint16) int {
+	for _, ar := range a {
+		if len(b) == 0 {
+			return 1
+		}
+		if int(ar) < len(upcase) {
+			ar = rune(upcase[int(ar)])
+		}
+		br := rune(binary.LittleEndian.Uint16(b))
+		bs := 2
+		if utf16.IsSurrogate(br) {
+			if len(b) == bs {
+				return 1 // error?
+			}
+			br = utf16.DecodeRune(br, rune(binary.LittleEndian.Uint16(b[bs:])))
+			if br == '\ufffd' {
+				return 1 // error?
+			}
+			bs += 2
+		} else {
+			br = rune(upcase[int(br)])
+		}
+		if ar < br {
+			return -1
+		} else if ar > br {
+			return 1
+		}
+		b = b[bs:]
+	}
+	if len(b) > 0 {
+		return -1
+	}
+	return 0
+}
+
+func bsearchLinkTable(b []byte, esize int, name string, upcase []uint16) ([]byte, error) {
+	if len(b) == 0 {
+		return nil, nil
+	}
+	n := binary.LittleEndian.Uint32(b[ltCountOff:])
+	es := b[ltEntryOff:]
+	nos := es[n*uint32(esize):]
```

Comment:
Most of the reader/writer code could use more descriptive variable names. This is giving me flashbacks to c haha